File: apps/inventory/formula_engine.py
"""
Expression-based formula engine for manifest standardization.

Syntax:
  Column references: [COLUMN_NAME]
  Functions: UPPER(expr), LOWER(expr), TITLE(expr), TRIM(expr),
             REPLACE(expr, find, replace), CONCAT(expr, ...), LEFT(expr, n), RIGHT(expr, n)
  Operators: + for string concatenation
  Literals: "quoted strings"

Examples:
  TITLE([Description])
  UPPER([Brand]) + " " + [Model]
  REPLACE([Category], "/", " - ")
  CONCAT(TRIM([Brand]), " ", TRIM([Model]))
"""
import re
from typing import Any


class FormulaError(Exception):
    pass


# ── Tokenizer ────────────────────────────────────────────────────────────────

TOKEN_PATTERNS = [
    ('STRING',    r'"(?:[^"\\]|\\.)*"'),
    ('COLREF',    r'\[([^\]]+)\]'),
    ('FUNC',      r'(UPPER|LOWER|TITLE|TRIM|REPLACE|CONCAT|LEFT|RIGHT)\s*(?=\()'),
    ('NUMBER',    r'\d+'),
    ('LPAREN',    r'\('),
    ('RPAREN',    r'\)'),
    ('COMMA',     r','),
    ('PLUS',      r'\+'),
    ('WS',        r'\s+'),
]

_TOKEN_RE = re.compile('|'.join(f'(?P<{name}>{pat})' for name, pat in TOKEN_PATTERNS))


def tokenize(formula: str) -> list[tuple[str, str]]:
    tokens = []
    pos = 0
    for m in _TOKEN_RE.finditer(formula):
        if m.start() != pos:
            bad = formula[pos:m.start()].strip()
            if bad:
                raise FormulaError(f"Unexpected characters: '{bad}' at position {pos}")
        pos = m.end()
        kind = m.lastgroup
        value = m.group()
        if kind == 'WS':
            continue
        if kind == 'STRING':
            value = value[1:-1].replace('\\"', '"').replace('\\\\', '\\')
        elif kind == 'COLREF':
            value = m.group('COLREF')
            value = value[1:-1]
        elif kind == 'FUNC':
            value = value.strip()
        tokens.append((kind, value))

    if pos != len(formula):
        remaining = formula[pos:].strip()
        if remaining:
            raise FormulaError(f"Unexpected characters at end: '{remaining}'")

    return tokens


# ── AST Nodes ────────────────────────────────────────────────────────────────

class ASTNode:
    pass

class StringLiteral(ASTNode):
    def __init__(self, value: str):
        self.value = value

class NumberLiteral(ASTNode):
    def __init__(self, value: int):
        self.value = value

class ColumnRef(ASTNode):
    def __init__(self, name: str):
        self.name = name

class FuncCall(ASTNode):
    def __init__(self, name: str, args: list[ASTNode]):
        self.name = name
        self.args = args

class ConcatOp(ASTNode):
    def __init__(self, left: ASTNode, right: ASTNode):
        self.left = left
        self.right = right


# ── Parser ───────────────────────────────────────────────────────────────────

class Parser:
    def __init__(self, tokens: list[tuple[str, str]]):
        self.tokens = tokens
        self.pos = 0

    def peek(self) -> tuple[str, str] | None:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def consume(self, expected_kind: str | None = None) -> tuple[str, str]:
        tok = self.peek()
        if tok is None:
            raise FormulaError("Unexpected end of expression")
        if expected_kind and tok[0] != expected_kind:
            raise FormulaError(f"Expected {expected_kind}, got {tok[0]} ('{tok[1]}')")
        self.pos += 1
        return tok

    def parse(self) -> ASTNode:
        node = self.parse_concat()
        if self.pos < len(self.tokens):
            tok = self.tokens[self.pos]
            raise FormulaError(f"Unexpected token: {tok[0]} ('{tok[1]}')")
        return node

    def parse_concat(self) -> ASTNode:
        left = self.parse_primary()
        while self.peek() and self.peek()[0] == 'PLUS':
            self.consume('PLUS')
            right = self.parse_primary()
            left = ConcatOp(left, right)
        return left

    def parse_primary(self) -> ASTNode:
        tok = self.peek()
        if tok is None:
            raise FormulaError("Unexpected end of expression")

        if tok[0] == 'STRING':
            self.consume()
            return StringLiteral(tok[1])

        if tok[0] == 'NUMBER':
            self.consume()
            return NumberLiteral(int(tok[1]))

        if tok[0] == 'COLREF':
            self.consume()
            return ColumnRef(tok[1])

        if tok[0] == 'FUNC':
            func_name = tok[1]
            self.consume()
            self.consume('LPAREN')
            args = []
            if self.peek() and self.peek()[0] != 'RPAREN':
                args.append(self.parse_concat())
                while self.peek() and self.peek()[0] == 'COMMA':
                    self.consume('COMMA')
                    args.append(self.parse_concat())
            self.consume('RPAREN')
            return FuncCall(func_name, args)

        if tok[0] == 'LPAREN':
            self.consume()
            node = self.parse_concat()
            self.consume('RPAREN')
            return node

        raise FormulaError(f"Unexpected token: {tok[0]} ('{tok[1]}')")


# ── Evaluator ────────────────────────────────────────────────────────────────

FUNCTIONS = {
    'UPPER':   lambda args: str(args[0]).upper(),
    'LOWER':   lambda args: str(args[0]).lower(),
    'TITLE':   lambda args: str(args[0]).title(),
    'TRIM':    lambda args: str(args[0]).strip(),
    'REPLACE': lambda args: str(args[0]).replace(str(args[1]), str(args[2])),
    'CONCAT':  lambda args: ''.join(str(a) for a in args),
    'LEFT':    lambda args: str(args[0])[:int(args[1])],
    'RIGHT':   lambda args: str(args[0])[-int(args[1]):] if int(args[1]) > 0 else '',
}

FUNC_ARG_COUNTS = {
    'UPPER': (1, 1),
    'LOWER': (1, 1),
    'TITLE': (1, 1),
    'TRIM':  (1, 1),
    'REPLACE': (3, 3),
    'CONCAT': (1, None),
    'LEFT':  (2, 2),
    'RIGHT': (2, 2),
}


def evaluate_ast(node: ASTNode, row: dict[str, str]) -> str:
    if isinstance(node, StringLiteral):
        return node.value

    if isinstance(node, NumberLiteral):
        return str(node.value)

    if isinstance(node, ColumnRef):
        return str(row.get(node.name, ''))

    if isinstance(node, ConcatOp):
        return evaluate_ast(node.left, row) + evaluate_ast(node.right, row)

    if isinstance(node, FuncCall):
        min_args, max_args = FUNC_ARG_COUNTS.get(node.name, (0, None))
        if len(node.args) < min_args:
            raise FormulaError(
                f"{node.name}() requires at least {min_args} argument(s), got {len(node.args)}"
            )
        if max_args is not None and len(node.args) > max_args:
            raise FormulaError(
                f"{node.name}() accepts at most {max_args} argument(s), got {len(node.args)}"
            )
        evaluated_args = [evaluate_ast(a, row) for a in node.args]
        return FUNCTIONS[node.name](evaluated_args)

    raise FormulaError(f"Unknown AST node type: {type(node).__name__}")


# ── Public API ───────────────────────────────────────────────────────────────

def evaluate_formula(formula: str, row: dict[str, str]) -> str:
    """Evaluate a formula expression against a row dict. Returns the result string.

    Raises FormulaError on syntax or evaluation errors.
    """
    formula = formula.strip()
    if not formula:
        return ''

    tokens = tokenize(formula)
    if not tokens:
        return ''

    parser = Parser(tokens)
    ast = parser.parse()
    return evaluate_ast(ast, row)


def validate_formula(formula: str) -> str | None:
    """Validate formula syntax without evaluating. Returns None if valid, error message if not."""
    try:
        formula = formula.strip()
        if not formula:
            return None
        tokens = tokenize(formula)
        if not tokens:
            return None
        parser = Parser(tokens)
        parser.parse()
        return None
    except FormulaError as e:
        return str(e)

File: frontend/src/components/inventory/preprocessing/formulaEngine.ts
/**
 * TS port of `apps/inventory/formula_engine.py` for client-side Sample Result + Formula Preview.
 */

export class FormulaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FormulaError';
  }
}

type TokKind =
  | 'STRING'
  | 'COLREF'
  | 'FUNC'
  | 'NUMBER'
  | 'LPAREN'
  | 'RPAREN'
  | 'COMMA'
  | 'PLUS';

type Token = { kind: TokKind; value: string };

/** Mirrors `apps/inventory/formula_engine.py` token patterns (sticky). */
const TOKEN_RE =
  /(?<STRING>"(?:[^"\\]|\\.)*")|(?<COLREF>\[([^\]]+)\])|(?<FUNC>(UPPER|LOWER|TITLE|TRIM|REPLACE|CONCAT|LEFT|RIGHT)\s*(?=\())|(?<NUMBER>\d+)|(?<LPAREN>\()|(?<RPAREN>\))|(?<COMMA>,)|(?<PLUS>\+)|(?<WS>\s+)/gy;

function tokenize(formula: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;
  while (pos < formula.length) {
    TOKEN_RE.lastIndex = pos;
    const m = TOKEN_RE.exec(formula);
    if (!m || m.index !== pos) {
      const bad = formula.slice(pos).trimStart().slice(0, 40);
      throw new FormulaError(`Unexpected characters: '${bad}' at position ${pos}`);
    }
    const g = m.groups as Record<string, string | undefined>;
    if (g.WS !== undefined) {
      pos = TOKEN_RE.lastIndex;
      continue;
    }
    let kind: TokKind;
    let value: string;
    if (g.STRING !== undefined) {
      kind = 'STRING';
      value = g.STRING.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
    } else if (g.COLREF !== undefined) {
      kind = 'COLREF';
      // Match the Python engine: strip the wrapping brackets from the full
      // COLREF token. Using a capture index here is fragile because earlier
      // alternation groups shift the numeric positions.
      value = g.COLREF.slice(1, -1);
    } else if (g.FUNC !== undefined) {
      kind = 'FUNC';
      value = g.FUNC.trim();
    } else if (g.NUMBER !== undefined) {
      kind = 'NUMBER';
      value = g.NUMBER;
    } else if (g.LPAREN !== undefined) {
      kind = 'LPAREN';
      value = '(';
    } else if (g.RPAREN !== undefined) {
      kind = 'RPAREN';
      value = ')';
    } else if (g.COMMA !== undefined) {
      kind = 'COMMA';
      value = ',';
    } else if (g.PLUS !== undefined) {
      kind = 'PLUS';
      value = '+';
    } else {
      throw new FormulaError('Tokenizer internal error');
    }
    tokens.push({ kind, value });
    pos = TOKEN_RE.lastIndex;
  }
  return tokens;
}

type ASTNode =
  | { type: 'str'; v: string }
  | { type: 'num'; v: number }
  | { type: 'col'; name: string }
  | { type: 'func'; name: string; args: ASTNode[] }
  | { type: 'concat'; left: ASTNode; right: ASTNode };

class Parser {
  private readonly tokens: Token[];
  private pos = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private peek(): Token | undefined {
    return this.tokens[this.pos];
  }

  private consume(expected?: TokKind): Token {
    const tok = this.peek();
    if (!tok) throw new FormulaError('Unexpected end of expression');
    if (expected && tok.kind !== expected) {
      throw new FormulaError(`Expected ${expected}, got ${tok.kind} ('${tok.value}')`);
    }
    this.pos += 1;
    return tok;
  }

  parse(): ASTNode {
    const node = this.parseConcat();
    if (this.pos < this.tokens.length) {
      const tok = this.tokens[this.pos];
      throw new FormulaError(`Unexpected token: ${tok.kind} ('${tok.value}')`);
    }
    return node;
  }

  private parseConcat(): ASTNode {
    let left = this.parsePrimary();
    while (this.peek()?.kind === 'PLUS') {
      this.consume('PLUS');
      const right = this.parsePrimary();
      left = { type: 'concat', left, right };
    }
    return left;
  }

  private parsePrimary(): ASTNode {
    const tok = this.peek();
    if (!tok) throw new FormulaError('Unexpected end of expression');

    if (tok.kind === 'STRING') {
      this.consume();
      return { type: 'str', v: tok.value };
    }
    if (tok.kind === 'NUMBER') {
      this.consume();
      return { type: 'num', v: Number.parseInt(tok.value, 10) };
    }
    if (tok.kind === 'COLREF') {
      this.consume();
      return { type: 'col', name: tok.value };
    }
    if (tok.kind === 'FUNC') {
      const name = tok.value;
      this.consume();
      this.consume('LPAREN');
      const args: ASTNode[] = [];
      if (this.peek()?.kind !== 'RPAREN') {
        args.push(this.parseConcat());
        while (this.peek()?.kind === 'COMMA') {
          this.consume('COMMA');
          args.push(this.parseConcat());
        }
      }
      this.consume('RPAREN');
      return { type: 'func', name, args };
    }
    if (tok.kind === 'LPAREN') {
      this.consume('LPAREN');
      const node = this.parseConcat();
      this.consume('RPAREN');
      return node;
    }

    throw new FormulaError(`Unexpected token: ${tok.kind} ('${tok.value}')`);
  }
}

const FUNCTIONS: Record<string, (args: string[]) => string> = {
  UPPER: (args) => String(args[0] ?? '').toUpperCase(),
  LOWER: (args) => String(args[0] ?? '').toLowerCase(),
  TITLE: (args) => titleCase(String(args[0] ?? '')),
  TRIM: (args) => String(args[0] ?? '').trim(),
  REPLACE: (args) =>
    String(args[0] ?? '').replace(String(args[1] ?? ''), String(args[2] ?? '')),
  CONCAT: (args) => args.map(String).join(''),
  LEFT: (args) => String(args[0] ?? '').slice(0, Number(args[1] ?? 0)),
  RIGHT: (args) => {
    const n = Number(args[1] ?? 0);
    const s = String(args[0] ?? '');
    return n > 0 ? s.slice(-n) : '';
  },
};

const FUNC_ARG_COUNTS: Record<string, [number, number | null]> = {
  UPPER: [1, 1],
  LOWER: [1, 1],
  TITLE: [1, 1],
  TRIM: [1, 1],
  REPLACE: [3, 3],
  CONCAT: [1, null],
  LEFT: [2, 2],
  RIGHT: [2, 2],
};

/** Match Python `str.title()` loosely for ASCII-ish manifests. */
function titleCase(s: string): string {
  return s.replace(/\w\S*/g, (w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase());
}

function evaluateAst(node: ASTNode, row: Record<string, string>): string {
  switch (node.type) {
    case 'str':
      return node.v;
    case 'num':
      return String(node.v);
    case 'col':
      if (import.meta.env.DEV) {
        const val = row[node.name];
        console.info(
          `[Preprocessing:S1][formulaEngine] column lookup ${JSON.stringify({
            extractedColName: node.name,
            colNameCharCodes: [...node.name].map((c) => c.charCodeAt(0)),
            rowHasKey: Object.prototype.hasOwnProperty.call(row, node.name),
            valueLen: val == null ? null : String(val).length,
            valueHead: val == null ? null : String(val).slice(0, 80),
          })}`,
        );
      }
      return row[node.name] ?? '';
    case 'concat':
      return evaluateAst(node.left, row) + evaluateAst(node.right, row);
    case 'func': {
      const [minA, maxA] = FUNC_ARG_COUNTS[node.name] ?? [0, 0];
      if (node.args.length < minA) {
        throw new FormulaError(`${node.name}() requires at least ${minA} argument(s), got ${node.args.length}`);
      }
      if (maxA != null && node.args.length > maxA) {
        throw new FormulaError(`${node.name}() accepts at most ${maxA} argument(s), got ${node.args.length}`);
      }
      const evaluated = node.args.map((a) => evaluateAst(a, row));
      const fn = FUNCTIONS[node.name];
      if (!fn) throw new FormulaError(`Unknown function ${node.name}`);
      return fn(evaluated);
    }
    default:
      throw new FormulaError('Unknown AST node');
  }
}

export function evaluateFormula(formula: string, row: Record<string, string>): string {
  const f = formula.trim();
  if (!f) return '';
  const tokens = tokenize(f);
  if (!tokens.length) return '';
  const parser = new Parser(tokens);
  const ast = parser.parse();
  return evaluateAst(ast, row);
}

export function evaluateFormulaSafe(formula: string, row: Record<string, string>): { ok: true; value: string } | { ok: false; error: string } {
  try {
    return { ok: true, value: evaluateFormula(formula, row) };
  } catch (e) {
    return { ok: false, error: e instanceof FormulaError ? e.message : String(e) };
  }
}
